### PR TITLE
User/authentication

### PR DESCRIPTION
**Pull request by barabas007**
Ногами не бейте.

### CODE REVIEW

File: .idea/.generators [hunk 1]
```diff
@@ -5,4 +5,4 @@ You are allowed to:
...
...eratorsGroup><Generator name="active_record:migration" /><Generator name="active_record:model" /><Generator name="active_record:observer" /><Generator name="active_record:session_migration" /><Generator name="controller" /><Generator name="erb:controller" /><Generator name="erb:mailer" /><Generator name="erb:scaffold" /><Generator name="generator" /><Generator name="helper" /><Generator name="integration_test" /><Generator name="mailer" /><Generator name="metal" /><Generator name="migration" /><Generator name="model" /><Generator name="model_subclass" /><Generator name="observer" /><Generator name="performance_test" /><Generator name="plugin" /><Generator name="resource" /><Generator name="scaffold" /><Generator name="scaffold_controller" /><Generator name="session_migration" /><Generator name="stylesheets" /><Generator name="test_unit:controller" /><Generator name="test_unit:helper" /><Generator name="test_unit:integration" /><Generator name="test_unit:mailer" /><Generator name="test_unit:model" /><Generator name="test_unit:observer" /><Generator name="test_unit:performance" /><Generator name="test_unit:plugin" /><Generator name="test_unit:scaffold" /></GeneratorsGroup></Settings>
```
Comment (alxekb) [inline .idea/.generators]: Please add this file to .gitignore

File: .idea/dataSources.xml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      <jdbc-driver>org.sqlite.JDBC</jdbc-driver>
+      <jdbc-url>jdbc:sqlite:$PROJECT_DIR$/db/development.sqlite3</jdbc-url>
+      <libraries>
+        <library>
+          <url>file://$APPLICATION_CONFIG_DIR$/jdbc-drivers/Xerial SQLiteJDBC/3.31.1/sqlite-jdbc-3.31.1.jar</url>
+        </library>
+        <library>
+          <url>file://$APPLICATION_CONFIG_DIR$/jdbc-drivers/Xerial SQLiteJDBC/3.31.1/license.txt</url>
+        </library>
+      </libraries>
```
Comment (alxekb) [inline .idea/dataSources.xml]: Here the same

File: .idea/railways.cache [hunk 3]
```diff
@@ -2,12 +2,15 @@ do /home/barabas/PROJECT/first_api/bin/bundle exec /usr/share/rvm/rubies/ruby-3.
...
-                                      PATCH  /users/:id(.:format)                                                                     users#update
-                                      PUT    /users/:id(.:format)                                                                     users#update
-                                      DELETE /users/:id(.:format)                                                                     users#destroy
+                         api_v1_login POST   /api/v1/login(.:format)                                                                  api/v1/login#create {:format=>:json}
+                         api_v1_users GET    /api/v1/users(.:format)                                                                  api/v1/users#show {:format=>:json}
+                                      PUT    /api/v1/users(.:format)                                                                  api/v1/users#edit {:format=>:json}
+                                      DELETE /api/v1/users(.:format)                                                                  api/v1/users#destroy {:format=>:json}
```
Comment (alxekb) [inline .idea/railways.cache]: Here the same thing

File: app/controllers/api/v1/login_controller.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+module Api
+  module V1
+    class LoginController < ApplicationController
+      rescue_from ActionController::ParameterMissing, with: :parameter_missing
+
+      def create
+        p params.require(:name).inspect
+        p params.require(:password).inspect
+        p params.require(:email).inspect
```
Comment (alxekb) [inline app/controllers/api/v1/login_controller.rb:9]: 😄  👍🏻

File: app/controllers/api/v1/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
+module Api
+  module V1
+    class UsersController < ApplicationController
+      def index; end
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb]:
```
def index
  render json: User.last(10)
end
```

File: app/controllers/api/v1/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+      def show
+        render json: User.all
+      end
+
+      def show_user
+        render json: User.limit(5).find(params[:id])
+      end
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb]: show_user -> show
```
render json: User.find_by(permitted_params[:id])
```

File: app/controllers/api/v1/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+        if user.save
+          render json: user, status: :ok
+        else
+          render json: user.errors, status: :unprocessable_entity
+        end
+      end
+
+      def destroy
+        User.find(params[:id]).destroy!
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb]: Please handle exceptions when use any bang method.

File: app/controllers/api/v1/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        else
+          render json: user.errors, status: :unprocessable_entity
+        end
+      end
+
+      def login
+        if User.find_by(email: params[:email])
+          render json: user, status: :ok
+        else
+          render json: user.errors, status: :unprocessable_entity
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb]: This won't work.

File: app/controllers/api/v1/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        end
+      end
+
+      def user_params
+        params.require(:user).permit(:name, :email, :password)
+      end
+
+
+
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb]: Please remove whitespace.

File: config/routes.rb [hunk 10]
```diff
@@ -1,3 +1,16 @@
+def method_user
+  post 'login', to: 'login#create'
+  get 'users', to: 'users#show'
+  put 'users', to: 'users#edit'
+  delete 'users', to: 'users#destroy'
+  get 'users', to: 'users#show_user'
+  post 'users', to: 'users#login'
+
+  resources :users, only: %i[show show_user create edit login]
+end
```
Comment (alxekb) [inline config/routes.rb]: 🙂

File: db/migrate/20210502181049_create_users.rb [hunk 11]
```diff
@@ -4,7 +4,7 @@ def change
       t.string :name
       t.string :email
       t.string :password
```
Comment (alxekb) [inline db/migrate/20210502181049_create_users.rb]: Please remove this.

File: spec/controllers/user_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+RSpec.describe UsersController, type: :controller do
+  let(:user) { create :user }
+
+  context 'POST #create' do
+    before { get :users }
+
+    it 'returns users' do
+      is_expected.to render_template :users
+      expect(assigns(:users)).to match_array(users)
+    end
```
Comment (alxekb) [inline spec/controllers/user_controller_spec.rb]: POST of get?
IMO this won't work.

File: spec/factories/user.rb [hunk 13]
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :user do
+    name       { Faker::Superhero.name }
+    email     { title.downcase }
+    password        { title.downcase }
+    #password_digest { title.downcase }
```
Comment (alxekb) [inline spec/factories/user.rb]: 👀

File: spec/reguests/login_spec.rb [hunk 14]
```diff
@@ -0,0 +1,27 @@
+require 'rails_helper'
+
+describe 'Login', type: :request do
+  describe 'POST/login' do
+    it 'login the client' do
+      post '/api/v1/login', params: { name: 'KKK', email: 'user@mail.ru', password: '123456' }
+
+      # expect(response).to have_http_status(:created)
+      # expect(response_body_if_short).to eq({ 'token' => '1234567890' })
```
Comment (alxekb) [inline spec/reguests/login_spec.rb]: 👀

File: spec/reguests/login_spec.rb [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    end
+
+    it 'returns error when email is missing' do
+      post '/api/v1/login', params: { name: 'KKK', password: '12345' }
+      expect(response).to have_http_status(:unprocessable_entity)
+    end
```
Comment (alxekb) [inline spec/reguests/login_spec.rb]: When you're using _when_ it's better to rewrite it with context block
```
context 'when something' do
  it 'returns an error' do
    # code here
  end
end
```

**Review by alxekb — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/login_controller.rb (hunk 4 above)
Comment (amveng) [reply app/controllers/api/v1/login_controller.rb:9]: наверно гем прай не указан в разрешенных )

File: db/migrate/20210502181049_create_users.rb (hunk 11 above)
Comment (barabas007) [reply db/migrate/20210502181049_create_users.rb]: я у себя все поля вижу в бд...там веток наплодилось...

File: app/controllers/api/v1/users_controller.rb [hunk 16]
```diff
@@ -1,7 +1,9 @@
 module Api
   module V1
     class UsersController < ApplicationController
-      def index; end
+      def index
+        render json: User.last(10)
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb]: Do we have any json serializer for this association?

File: app/controllers/api/v1/users_controller.rb [hunk 17]
```diff
@@ -14,16 +16,19 @@ def show_user
       def create
         user = User.new(user_params)
 
-        if user.save
-          render json: user, status: :ok
+        if user.persisted?
+          render json: user, status: :created
         else
           render json: user.errors, status: :unprocessable_entity
         end
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb]: In this case user won't be saved at all

**Review by alxekb — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/users_controller.rb (hunk 16 above)
Comment (barabas007) [reply app/controllers/api/v1/users_controller.rb]: You mean a model?

File: app/controllers/api/v1/users_controller.rb [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+        if User.find_by(id: params[:id])&.destroy
+          render json: { success: true }
+        else
+          render json: { user: 'not found' }, status: :unprocessable_entity
+        end
+      end
+
+      def update
+        user = User.find_by(id: params[:id])
+        if user
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb:33]:
```
ApiErrorsHandler.not_found user_params if user.empty?

if user.update(user_params)
  render json ...
```

File: app/controllers/api/v1/users_controller.rb [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+        user = User.find_by(email: params[:email])
+        if user && user.authenticate(params[:password])
+          render json: UserBlueprint.render(user), status: :ok     #{ success: true }
+        else
+          render json: { user: 'not found' }, status: 400
+        end
+      end
+
+      def user_params
+        params.permit(:name, :email, :password, :password_confirmation)
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb:51]:
```
params.require(:user).permit(:id, ... )
```

File: app/controllers/api/v1/users_controller.rb [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+          user.update(user_params)
+          render json: UserBlueprint.render(user), status: :ok  #{ success: true },
+        else
+          render json: {  user: 'forbidden' }, status: :forbidden
+        end
+      end
+
+      def login
+        user = User.find_by(email: params[:email])
+        if user && user.authenticate(params[:password])
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb:43]: I can't find definition of the `#authenticate` method.

File: app/controllers/api/v1/users_controller.rb [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+        if user
+          user.update(user_params)
+          render json: UserBlueprint.render(user), status: :ok  #{ success: true },
+        else
+          render json: {  user: 'forbidden' }, status: :forbidden
+        end
+      end
+
+      def login
+        user = User.find_by(email: params[:email])
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb:42]:
```
User.find_by(email: user_params[:email])
```

File: app/controllers/api/v1/users_controller.rb [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+        user = User.new(user_params)
+        if user.save
+          render json: UserBlueprint.render(user)   #{ success: true, user: user.as_json }
+        else
+          render json: { user: 'not unauthorized' }, status: :unauthorized
+        end
+      end
+
+      def destroy
+        if User.find_by(id: params[:id])&.destroy
```
Comment (alxekb) [inline app/controllers/api/v1/users_controller.rb:24]: Please use permitted params.

**Review by alxekb — CHANGES_REQUESTED** [review]

**Comment by barabas007** [comment]
123

**Pull request merged**